source/MaterialXGenGlsl/Nodes/ConvolutionGLSL.cpp

```diff
@@ -0,0 +1,137 @@
+#include <MaterialXGenGlsl/Nodes/ConvolutionGlsl.h>
+#include <MaterialXGenShader/HwShader.h>
+#include <MaterialXGenShader/ShaderGenerator.h>
+
+namespace MaterialX
+{
+ConvolutionGlsl::ConvolutionGlsl()
+    : _sampleCount(1)
+    , _filterSize(1.0f)
+    , _filterOffset(0.0f)
+{
+}
+
+void ConvolutionGlsl::emitInputSamplesUV(const SgNode& node, SgNodeContext& context, ShaderGenerator& shadergen, HwShader& shader)
+{
+    const SgInput* inInput = node.getInput("in");
+    string inputName = inInput->name;
+
+    // Require an upstream node to sample
+    SgNode* upstreamNode = nullptr;
+    SgOutput* inConnection = inInput->connection;
+
+    if (inConnection && inConnection->type && acceptsInputType(inConnection->type))
+    {
+        upstreamNode = inConnection->node;
+        if (upstreamNode && upstreamNode->hasClassification(SgNode::Classification::SAMPLE2D))
+        {
+            SgImplementation *impl = upstreamNode->getImplementation();
+            if (impl)
+            {
+                SgOutput* upstreamOutput = upstreamNode->getOutput();
+                if (upstreamOutput)
+                {
+                    string outputName = upstreamOutput->name;
+
+                    // Find out which input needs to be sampled multiple times
+                    SgInput* samplingInput = upstreamNode->getSamplingInput();
+                    // TODO: For now we only support uv space sampling
+                    if (samplingInput && samplingInput->type != Type::VECTOR2)
+                    {
+                        samplingInput = nullptr;
+                    }
+
+                    if (samplingInput)
+                    {
+                        // This is not exposed. Assume a filter size of 1 with no offset
+
+                        // Emit code to compute sample size
+                        //
+                        string sampleInputValue;
+                        shadergen.getInput(context, samplingInput, sampleInputValue);
+
+                        const string sampleSizeName(node.getOutput()->name + "_sample_size");
+                        string sampleCall("vec2 " + sampleSizeName + " = " +
+                            _sampleSizeFunctionUV + "(" +
+                            sampleInputValue + "," +
+                            std::to_string(_filterSize) + "," +
+                            std::to_string(_filterOffset) + ");"
+                        );
+                        shader.addLine(sampleCall);
+
+                        // Build the sample offset strings. This is dependent on
+                        // the derived class to determine where samples are located
+                        // and to generate the strings required to offset from the center
+                        // sample. The sample size is passed over.
+                        //
+                        StringVec inputVec2Suffix;
+                        computeSampleOffsetStrings(sampleSizeName, inputVec2Suffix);
```

Comment:
Derived convolution classes provide the offset strings. They are not interpreted here. i.e. The base class doesn't know the kernel shape nor what each element corresponds to.